### PR TITLE
Adding subscriptions resource implementation

### PR DESCRIPTION
**Pull request by malinthaprasan**

### CODE REVIEW

**Comment by hevayo** [comment]
This was manually merged

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java [hunk 1]
```diff
@@ -1,42 +1,115 @@
...
+                //when retrieving subscriptions from getSubscribedAPIs(), existing application ids are not populated
+                if (!StringUtils.isEmpty(applicationId)) {
+                    subscriptionDTO.setApplicationId(applicationId);
+                }
+                subscriptionDTOs.add(subscriptionDTO);
+            }
+            return Response.ok().entity(subscriptionDTOs).build();
+
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java]: Use proper exception handling.

Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java]: Fixed in #1303 (https://github.com/wso2/carbon-apimgt/commit/8049b7e2779abb3dd6bf13aeb70b062438e7c5b7#diff-c7ba6d5407040021b28a360a230e2981R143)

Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java]: Fixed in #1303 (https://github.com/wso2/carbon-apimgt/commit/8049b7e2779abb3dd6bf13aeb70b062438e7c5b7) and #1533 (https://github.com/wso2/carbon-apimgt/commit/243966339e63e60a915cb07ee4e5562a9cb7facf)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java [hunk 2]
```diff
@@ -1,42 +1,115 @@
...
+            APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifier(apiId);
+            apiIdentifier.setTier(body.getTier());
+            int subscriptionId =
+                    apiConsumer.addSubscription(apiIdentifier, username, Integer.parseInt(applicationId));
+            SubscribedAPI addedSubscribedAPI = apiConsumer.getSubscriptionById(subscriptionId);
+            SubscriptionDTO addedSubscriptionDTO = SubscriptionMappingUtil.fromSubscriptiontoDTO(addedSubscribedAPI);
+            //todo: use a proper way other than using "subscriptions/"
+            return Response.created(new URI("subscriptions/" + subscriptionId)).entity(addedSubscriptionDTO).build();
+        } catch (APIManagementException | URISyntaxException e) {
+            throw new InternalServerErrorException(e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java]: Exception handling already fixed in #1303 (https://github.com/wso2/carbon-apimgt/commit/8049b7e2779abb3dd6bf13aeb70b062438e7c5b7) and #1533 (https://github.com/wso2/carbon-apimgt/commit/243966339e63e60a915cb07ee4e5562a9cb7facf)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java [hunk 3]
```diff
@@ -1,42 +1,115 @@
...
-        return Response.ok().entity(new ApiResponseMessage(ApiResponseMessage.OK, "magic!")).build();
+    public Response subscriptionsSubscriptionIdDelete(String subscriptionId, String ifMatch, String ifUnmodifiedSince) {
+        String username = RestApiUtil.getLoggedInUsername();
+        APIConsumer apiConsumer = null;
+        try {
+            apiConsumer = RestApiUtil.getConsumer(username);
+            apiConsumer.removeSubscriptionById(Integer.parseInt(subscriptionId));
+            return Response.ok().build();
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java]: Exception handling already fixed in #1303 (https://github.com/wso2/carbon-apimgt/commit/8049b7e2779abb3dd6bf13aeb70b062438e7c5b7) and #1533 (https://github.com/wso2/carbon-apimgt/commit/243966339e63e60a915cb07ee4e5562a9cb7facf)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java [hunk 4]
```diff
@@ -1,42 +1,115 @@
...
+            apiConsumer = RestApiUtil.getConsumer(username);
+            SubscribedAPI subscribedAPI = apiConsumer.getSubscriptionById(Integer.parseInt(subscriptionId));
+            if (subscribedAPI != null) {
+                SubscriptionDTO subscriptionDTO = SubscriptionMappingUtil.fromSubscriptiontoDTO(subscribedAPI);
+                return Response.ok().entity(subscriptionDTO).build();
+            } else {
+                throw new NotFoundException();
+            }
+        } catch (APIManagementException e) {
+            throw new InternalServerErrorException(e);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/SubscriptionsApiServiceImpl.java]: Exception handling already fixed in #1303 (https://github.com/wso2/carbon-apimgt/commit/8049b7e2779abb3dd6bf13aeb70b062438e7c5b7) and #1533 (https://github.com/wso2/carbon-apimgt/commit/243966339e63e60a915cb07ee4e5562a9cb7facf)

**Pull request closed without merging**